### DISCUSSION TITLE
Proposal: Replace strmangle with flect

### DISCUSSION THREAD

**Discussion by ccakes** [aarondl/sqlboiler#1092] [Ideas]
Per the contributing guidelines, raising an issue to discuss this before I start working on an implementation.

`strmangle` is a bit difficult to work with, it lacks a public API for modifying the predefined `boilRuleset` which makes it impossible for projects embeding sqlboiler to define their own acronyms and inflection rules (eg for industry-specific terms). `strmangle` looks like a simple wrapper around [volatiletech/inflect](https://github.com/volatiletech/inflect) which itself looks like a fork of [markbates/inflect](https://github.com/markbates/inflect).

If you'd be open to replacing this machinery, [gobuffalo/flect](https://github.com/gobuffalo/flect) is a more modern inflection engine designed to replace the markbates package. I haven't looked into this **at all** but am happy to if a successful implementation is likely to be merged.

If I went ahead, the goals would be (roughly..):
- Functionally-compatible with sqlboiler@v4 - ie drop in replacement where all v4 tests pass.
    - If unable to make this work, then could target v5
- Expose functionality to library users to extend/modify the internal inflection rules

Thoughts?

**Comment by stephenafamo** [comment]
I'm not sure we have tests to check if `strmangle` works **a specific** way, so I would say if we are changing this package we just target v5 directly. Because if the inflection is different, then the generated code will no longer be backward compatible and we don't want that.

Personally, I am all for using a well-maintained 3rd party package instead of something that we have to maintain ourselves. However if we want to do this, we'd have to make sure that `flect` is the best choice (at least at the moment). Are there any other alternatives?

**Reply by stephenafamo** [reply]
So, I took some time to investigate this today. 

### The limitation of `strmangle`

The main issue you raised with `strmangle` was that it was difficult for users to add to the ruleset. While it is possible with `flect`, it doesn't matter since SQLBoiler runs as a CLI tool. We could add configuration options that help the user define more inflections, and by extension, we could add the necessary functions directly to `strmangle` to achieve this.

### The performance of `flect`

While `flect` does work quite well some rudimentary benchmarks show it to be many many times slower than `strmangle`.

For title casing, it is 50 to 200 times slower in many cases, and for singular/plural inflection, it is about 1.4x slower for alphanumeric strings, and gets up to 5x slower once spaces or symbols are present.

Since there is no clear benefit of switching to `flect`, I'd say the bad performance is a deal breaker here.

**Reply by ccakes** [reply]
Sorry for the slow reply - based on performance I think your assessment makes sense.

> While it is possible with flect, it doesn't matter since SQLBoiler runs as a CLI tool.

It is possible to call sqlboiler as a library which is what I'm doing, and tbh it works pretty well. The only frustrating hard boundary where a CLI is called is at the driver handoff.
